syntax = "proto3";

package raft;

import "google/protobuf/empty.proto";

service Raft {
    rpc SendVoteRequest(VoteRequest) returns (google.protobuf.Empty);
    rpc SendVoteResponse(VoteResponse) returns (google.protobuf.Empty);
    rpc SendClientRequest(ClientRequest) returns (ClientResponse);
    rpc SendLogRequest(LogRequest) returns (google.protobuf.Empty);
    rpc SendLogResponse(LogResponse) returns (google.protobuf.Empty);
}

message ClientRequest {
    string RequestType = 1;  // "GET", "SET", "DELETE"
    string Key = 2;
    string Value = 3;
}

message ClientResponse {
    string Key = 1;
    string Value = 2;
    string Status = 3;
}

message LogEntry {
    int64 Term = 1;
    string Msg = 2;
}

message VoteRequest {
    string FromNode = 1;
    int64 Term = 2;
    int64 LogLength = 3;
    int64 LogTerm = 4;
}

message VoteResponse {
    string FromNode = 1;
    int64 Term = 2;
    bool Granted = 3;
}

message LogRequest {
    string LeaderId = 1;
    int64 Term = 2;
    int64 LogLength = 3;
    int64 LogTerm = 4;
    int64 LeaderCommit = 5;
    repeated LogEntry Entries = 6;
}

message LogResponse {
    string Follower = 1;
    int64 Term = 2;
    int64 Ack = 3;
    bool Success = 4;
}
